### DISCUSSION TITLE
The codeberg mirror is not up to date

### DISCUSSION THREAD

**Discussion by theoparis** [GitoxideLabs/gitoxide#1487] [General]
See https://codeberg.org/Byron/gitoxide, last commit was 2 months ago 🤔 
Maybe this needs to be automated.

**Comment by Byron** [comment]
Thanks for the hint!

Indeeds it's a manual process and I usually do it after a major feature was merged - this didn't happen in way too long.

Also, when trying to push now, I get this:

```
❯ gp codeberg main
Enumerating objects: 3578, done.
Counting objects: 100% (3454/3454), done.
Delta compression using up to 10 threads
Compressing objects: 100% (1333/1333), done.
send-pack: unexpected disconnect while reading sideband packet
Writing objects: 100% (2650/2650), 3.43 MiB | 30.52 MiB/s, done.
Total 2650 (delta 1952), reused 1831 (delta 1254), pack-reused 0
fatal: the remote end hung up unexpectedly
Everything up-to-date
```

And despite wonky error handling, it doesn't actually work and the remote is not uptodate.

I wish `gitoxide` could already do that.

**Reply by EliahKagan** [reply]
Now that several days have passed, I'm curious if the `git push` would just succeed now. If not, I'm curious if changing from SSH to HTTPS, or from HTTPS to SSH, would let it work.

I was able to make a fork of gitoxide on Codeberg and to push the same commits [to a branch on my fork](https://codeberg.org/EliahKagan/gitoxide/src/branch/gh). I used SSH. This suggests that using SSH might work if HTTPS is not working, though of course other factors such as region and the exact software being used could be involved.

Another approach, though, is that I could open a PR on Codeberg proposing that those commits be merged into your (i.e. the upstream) repository there. The PR [would look like this](https://codeberg.org/EliahKagan/gitoxide/pulls/1) but would be on your repo rather than on my fork for testing.

Unlike GitHub, Codeberg facilitates fast-forward merges from within its web-based interface, so this is another way to get the main branch there to be an up-to-date mirror of the main branch here.

> ![Screenshot of pull request page on Codeberg showing the menu for selecting ways to merge a PR, emphasizing the "Fast-forward only" option](https://github.com/user-attachments/assets/128c2e4a-30ab-4de1-9502-301898c58b64)

For the longer term, if the reason pushes to the Codeberg mirror have not been automated is to prevent security breaches in access to this repository from giving an attacker access there as well, then maybe having a less privileged account automatically open PRs (perhaps on a schedule and also in response to particular events) with `berg pull ...`, which could be manually looked over and then accepted with fast-forward merges, would be a reasonable middle ground.

**Reply by Byron** [reply]
That's interesting! For me it still doesn't work via HTTPS and SSH I didn't setup.
~~Somehow, I don't really want to either and rather wait until it maybe works again a year from now.~~

Interestingly, I also pushed to `gitee`, and it has the same issue.
Using a much more recent version of `git` fixes the issue. That's a strange thing to see, Git is usually very mindful about breaking backwards compatibility.

```
❯ PATH=~/dev/github.com/git/git ~/dev/github.com/git/git/git push gitee main
Username for 'https://gitee.com': byronbates
Password for 'https://byronbates@gitee.com':
Enumerating objects: 3874, done.
Counting objects: 100% (3749/3749), done.
Delta compression using up to 10 threads
Compressing objects: 100% (1493/1493), done.
Writing objects: 100% (2926/2926), 3.49 MiB | 1.66 MiB/s, done.
Total 2926 (delta 2163), reused 1974 (delta 1370), pack-reused 0 (from 0)
remote: Resolving deltas: 100% (2163/2163), completed with 387 local objects.
remote: Powered by GITEE.COM [1.1.5]
remote: Set trace flag a2476185
To https://gitee.com/byronbates/gitoxide.git
   2f9f0ac36..7dff44754  main -> main

gitoxide ( main) [$] took 11s
❯ PATH=~/dev/github.com/git/git ~/dev/github.com/git/git/git push codeberg main
Username for 'https://codeberg.org': Byron
Password for 'https://Byron@codeberg.org':
Enumerating objects: 3874, done.
Counting objects: 100% (3749/3749), done.
Delta compression using up to 10 threads
Compressing objects: 100% (1493/1493), done.
Writing objects: 100% (2926/2926), 3.49 MiB | 4.35 MiB/s, done.
Total 2926 (delta 2163), reused 1974 (delta 1370), pack-reused 0 (from 0)
remote: Resolving deltas: 100% (2163/2163), completed with 387 local objects.
To https://codeberg.org/Byron/gitoxide.git
   2f9f0ac36..7dff44754  main -> main
```

For posterity, the Git version I am using is `git version 2.39.3 (Apple Git-146)`.
